### PR TITLE
Streaming ChatGPT example

### PR DESCRIPTION
**Pull request by neilberkman**
https://user-images.githubusercontent.com/37657/224513447-d73895c7-0865-4b3a-b7d6-d30b498a978c.mp4

Not sure if this fits the repo based on your comment [here](https://github.com/wojtekmach/mix_install_examples/pull/4#issuecomment-934154383) but I thought it made for an interesting demo.

### CODE REVIEW

File: chatgpt_streaming.exs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+  def handle_info(
+        %EventsourceEx.Message{
+          data: jason_payload,
+          event: "message"
+        },
+        state
+      ) do
+    message = Jason.decode!(jason_payload)
+    extract_text(message) |> IO.write()
```
Comment (wojtekmach) [inline chatgpt_streaming.exs]:
```suggestion
    jason_payload |> Jason.decode!() |> extract_text() |> IO.write()
```

File: chatgpt_streaming.exs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    {:noreply, state}
+  end
+
+  def extract_text(%{"choices" => choices}) do
+    choices
+    |> Enum.map(fn %{"delta" => delta} -> delta end)
+    |> Enum.filter(fn content -> Map.has_key?(content, "content") end)
+    |> Enum.map(fn %{"content" => content} -> content end)
+    |> Enum.join()
```
Comment (wojtekmach) [inline chatgpt_streaming.exs]: I believe this is equivalent to the following which would do everything in a single pass:

```suggestion
    for %{"delta" => delta} <- choices,
        %{"content" => content} <- delta,
        into: "" do
      content
    end
```

could you double-check?

**Review by wojtekmach — COMMENTED** [review]
Thank you for the PR! Could you please rename the file to: `eventsource_ex_chatgpt`?

File: chatgpt_streaming.exs (hunk 2 above)
Comment (neilberkman) [reply chatgpt_streaming.exs]: Thanks @wojtekmach! That works with the addition of a filter expression:

https://github.com/wojtekmach/mix_install_examples/pull/22/commits/b937d00475116dd4369a34c1a668e15eefd593da#diff-42fd847355cc16aa5adf379abaeb4d43225cd32098a9d7f565fae8206a061042R45-R49

File: eventsource_ex_chatgpt.exs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        state
+      ) do
+    jason_payload |> Jason.decode!() |> extract_text() |> IO.write()
+
+    {:noreply, state}
+  end
+
+  def extract_text(%{"choices" => choices}) do
+    for %{"delta" => delta} <- choices,
+        %{"content" => content} <- if(Map.has_key?(delta, "content"), do: [delta], else: []),
```
Comment (wojtekmach) [inline eventsource_ex_chatgpt.exs]: oh if `delta` is a map, let's go with:

```elixir
for %{"delta" => %{"content" => content}} <- choices, into: "", do: content
```

:)

**Comment by wojtekmach** [comment]
Thank you!

**Pull request merged**